r/R/install-arrow.R

```diff
@@ -137,6 +137,262 @@ reload_arrow <- function() {
+#   the filename and URL array
+# @return A list with two character vectors, with names `filenames` and `urls`
+#
+# The output of this function has split out the filename and URL components,
+# but has not yet substituted in the version numbers. The output is next passed
+# to `..install_substitute_all()`
+#
+# Used as a helper to parse versions.txt
+..install_parse_dependency_array <- function(array_lines) {
+  stopifnot(
+    length(array_lines) >= 1,
+    is.character(array_lines),
+    !anyNA(array_lines)
+  )
+  array_lines <- trimws(array_lines)
+
+  # Parse the array_lines with a regex. Each line of the array is a different
+  # component, with a format like `"<component name> <filename> <url>"` (quotes
+  # included). The filename and URL include some version string that's defined
+  # earlier in the file.
+  # Regex in words:
+  # Start with `"ARROW_`, then any capital ASCII letter, number, or underscore.
+  # After a space, find anything except a space, colon, or forward slash. (No
+  # space is essential, and would be essential to parsing the array in bash.
+  # The colon and slash are just basic guards that this is a filename.) Next, a
+  # space. Then a URL, starting with https://, and including anything except a
+  # space. (This is the URL before substituting in the version sting, so normal
+  # URL parsing rules don't apply.)
+  dep_array_regex <- '^"(ARROW_[A-Z0-9_]+_URL) ([^ :/"]+) (https://[^ "]+)"$'
+  if (!all(grepl(dep_array_regex, array_lines, perl = TRUE))) {
+    stop("Cannot parse thirdparty dependency array in expected format.")
+  }
+  list(
+    filenames = gsub(dep_array_regex, "\\2", array_lines, perl = TRUE),
+    urls      = gsub(dep_array_regex, "\\3", array_lines, perl = TRUE)
+  )
+}
+
+# Parse the versions.txt file
+#
+# @param versions_file Filename pointing to versions.txt
+# @return The parsed and ready-to-use values, as a named list of vectors
+#
+#
+# The versions.txt file is included as part of the R tar file, and is here:
+# https://github.com/apache/arrow/blob/master/cpp/thirdparty/versions.txt
+#
+# Used as a helper to parse versions.txt
+..install_parse_lines <- function(versions_file) {
+  orig_lines <- readLines(versions_file)
+
+  lines <- gsub("#.*", "", orig_lines, perl = TRUE)
+  lines <- lines[lines != ""]
+
+  dep_array_start_idx <- grep("^DEPENDENCIES=\\($", lines, perl = TRUE)
+  dep_array_lines <- lines[
+    seq.int(from = dep_array_start_idx + 1, to = length(lines) - 1, by = 1)
+  ]
+  version_lines <- lines[seq.int(1, dep_array_start_idx - 1, by = 1)]
+  version_info <- ..install_parse_version_lines(version_lines)
+
+  failed_to_parse <- c(
+    anyNA(orig_lines),
+    length(orig_lines) > 1000,
+    length(lines) == 0,
+    length(dep_array_start_idx) != 1,
+    dep_array_start_idx <= 1,
+    dep_array_start_idx >= length(lines) - 3,
+    lines[length(lines)] != ")",
+    length(dep_array_lines) == 0,
+    anyNA(version_info)
+  )
+
+  if (any(failed_to_parse)) {
+    stop(
```

Comment:
You could use `rlang::abort` here but that is very heterogenous in the package so 🤷 (and you aimed for base R which is good!)